auth/groups.go

```diff
@@ -0,0 +1,177 @@
+package auth
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+const MaxLevel = uint64(5)
+const MinLevel = uint64(1)
+
+var (
+	// ErrMaxLevelExceeded malformed entity.
+	ErrMaxLevelExceeded = errors.New("level must be less than or equal 5")
+
+	// ErrBadGroupName malformed entity.
+	ErrBadGroupName = errors.New("incorrect group name")
+
+	// ErrGroupConflict group conflict.
+	ErrGroupConflict = errors.New("group already exists")
+
+	// ErrCreateGroup indicates failure to create group.
+	ErrCreateGroup = errors.New("failed to create group")
+
+	// ErrFetchGroups indicates failure to fetch groups.
+	ErrFetchGroups = errors.New("failed to fetch groups")
+
+	// ErrUpdateGroup indicates failure to update group.
+	ErrUpdateGroup = errors.New("failed to update group")
+
+	// ErrDeleteGroup indicates failure to delete group.
+	ErrDeleteGroup = errors.New("failed to delete group")
+
+	// ErrGroupNotFound indicates failure to find group.
+	ErrGroupNotFound = errors.New("failed to find group")
+
+	// ErrAssignToGroup indicates failure to assign member to a group.
+	ErrAssignToGroup = errors.New("failed to assign member to a group")
+
+	// ErrUnassignFromGroup indicates failure to unassign member from a group.
+	ErrUnassignFromGroup = errors.New("failed to unassign member from a group")
+
+	// ErrUnsupportedContentType indicates unacceptable or lack of Content-Type
+	ErrUnsupportedContentType = errors.New("unsupported content type")
+
+	// ErrFailedDecode indicates failed to decode request body
+	ErrFailedDecode = errors.New("failed to decode request body")
+
+	// ErrMissingParent indicates that parent can't be found
+	ErrMissingParent = errors.New("failed to retrieve parent")
+
+	// ErrGroupNotEmpty indicates group is not empty, can't be deleted.
+	ErrGroupNotEmpty = errors.New("group is not empty")
+
+	// ErrMemberAlreadyAssigned indicates that members is already assigned.
+	ErrMemberAlreadyAssigned = errors.New("member is already assigned")
+
+	// ErrSelectEntity indicates error while reading entity from database
+	ErrSelectEntity = errors.New("select entity from db error")
+)
+
+type GroupMetadata map[string]interface{}
+
+type Member struct {
+	ID   string
+	Type string
+}
+
+type Group struct {
+	ID          string
+	OwnerID     string
+	ParentID    string
+	Name        string
+	Description string
+	Metadata    GroupMetadata
+	// Indicates a level in tree hierarchy.
+	// Root node is level 1.
+	Level int
+	// Path in a tree consisting of group ids
+	// parentID1.parentID2.childID1
+	// e.g. 01EXPM5Z8HRGFAEWTETR1X1441.01EXPKW2TVK74S5NWQ979VJ4PJ.01EXPKW2TVK74S5NWQ979VJ4PJ .
```

Comment:
For me the full stop at the end of a sentence caused for a moment a bit of confusion because there are also dots in this tree structure. I would remove `.` at end.